src/Grpc.Net.Client/Internal/GrpcCall.cs

```diff
@@ -157,35 +159,61 @@ private void DisposeCore()
         {
             if (!Disposed)
             {
-                Disposed = true;
-
-                if (!ResponseFinished)
-                {
-                    // If the response is not finished then cancel any pending actions:
-                    // 1. Call HttpClient.SendAsync
-                    // 2. Response Stream.ReadAsync
-                    // 3. Client stream
-                    //    - Getting the Stream from the Request.HttpContent
-                    //    - Holding the Request.HttpContent.SerializeToStream open
-                    //    - Writing to the client stream
-                    CancelCall();
-                }
-                else
+                lock (this)
                 {
-                    _writeStreamTcs?.TrySetCanceled();
-                    _completeTcs?.TrySetCanceled();
+                    if (!Disposed)
+                    {
+                        Disposed = true;
+
+                        var resolvedStatus = _responseStatus ?? StatusCode.Cancelled;
+                        if (resolvedStatus != StatusCode.OK)
+                        {
+                            GrpcEventSource.Log.CallFailed(resolvedStatus);
+                        }
+                        GrpcEventSource.Log.CallStop();
+
+                        if (_activity != null)
```

Comment:
do you know if  Dispose (i.e. Activity.Stop) happens in the same execution context as Start?

Basically, if someone does this:

```csharp
  GrpcCall call = await Task.Run (() => return GrpcCall());
  await call .SendTask;
  call.Dispose();
```

will cause a bugs that are very hard to find:
- `Task.Run (() => return GrpcCall());` will create activity that will only exists inside this task
- `Dispose` will attempt to stop the instance of this activity, but listener (tracer) will attempt to get info from `Activity.Current` - it doesn't know which instance of activity is stopped and always deals with current.

another one:
if when grpc call is created, Current was A and when it was Disposed it was B, after dispose Current will turn to A
```csharp
  var a = new Actitvity("a").Start();
  GrpcCall call = await Task.Run (() => return GrpcCall());
  var res = await call.SendTask;
  var b = new Actitvity("b").Start();
  call.Dispose();

  Console.WriteLine(Activity.Current.OperationName); // you'd expect it to be 'b', but it's 'a'
```

It seems to affect logger scope in the same way (some loggers put scopes in AsyncLocal). I don't know if it matters or may be this API is intended to be used in only one way that never allows this. 
If not, I suggest to move activity stop to the end of `SendAsync` - this will guarantee that it's actually the same execution path or at least leave a comment somewhere that it's a danger zone.

Reply:
I don't think SendAsync will work. The lifetime of a gRPC isn't start request -> get response. A gRPC call ends in one of these situations:

1. The response is received and http status or headers are incorrect
2. Error when sending or receiving message
3. Deadline timer executes
4. User calls Dispose on call object
5. The response body has finished reading

The deadline timer is definitely on a different thread. All the others ends could happen on a different thread.

Reply:
different threads and callbacks are not necessarily a problem as long as they are async-friendly. AsyncLocal flows through async calls and doesn't care about threads.
But it seems it's not the case here.

There may be some tricks with execution context (capture and restore it), but I'm no expert and don't know if it is a good idea. Maybe @davidfowl knows who can help with execution context?



The ~~hacky~~ ways I see here are:

In dispose, check if _activity == Current. If it is - we are good (but it seems not very likely- who knows where user disposes calls)
If it's not current, 
 * keep old current value in local var
 * set Activity.Current  to _activity (bummer, public Current setter comes with DiagnosticSource 4.6.0)
 * After Stop completes return old current back

Another way: Put _activity instance into Stop event payload and expect listener do a special treatment (know the activity.current does not work). I don't like it as it additional contract and if listener doesn't know about it, it leads to hard-to-debug issues. Also, in general, we want to have generic listeners that don't know anything about payloads and subscribe to all stuff in the same way.

Reply:
I've refactored code to stop the activity via CancellationToken.Register. That will capture the execution context.

A test of your situation passes.

Reply:
Refactored to use a TCS instead of CT.Register